aptos-move/framework/aptos-framework/sources/storage_gas.move

```diff
@@ -0,0 +1,349 @@
+        *borrow_global_mut<StorageGasConfig>(@aptos_framework) = config;
+    }
+
+    public(friend) fun initialize(aptos_framework: &signer) {
+        system_addresses::assert_aptos_framework(aptos_framework);
+        assert!(
+            !exists<StorageGasConfig>(@aptos_framework),
+            error::already_exists(ESTORAGE_GAS_CONFIG)
+        );
+        let item_curve = base_32_exponential_curve(100, 10000);
+        let byte_curve = base_32_exponential_curve(1, 1000);
+
+        let item_config = UsageGasConfig {
+            target_usage: 1000000000,  // 1 billion
+            read_curve: copy item_curve,
+            create_curve: copy item_curve,
+            write_curve: copy item_curve,
+        };
+        let byte_config = UsageGasConfig {
+            target_usage: 250000000000, // 250 GB
+            read_curve: copy byte_curve,
+            create_curve: copy byte_curve,
+            write_curve: copy byte_curve,
+        };
+        validate_usage_config(&item_config);
+        validate_usage_config(&item_config);
+        move_to(aptos_framework, StorageGasConfig {
+            item_config,
+            byte_config,
+        });
+
+        assert!(
+            !exists<StorageGas>(@aptos_framework),
+            error::already_exists(ESTORAGE_GAS)
+        );
+        move_to(aptos_framework, StorageGas {
+            per_item_read: 0,
+            per_item_create: 0,
+            per_item_write: 0,
+            per_byte_read: 0,
+            per_byte_create: 0,
+            per_byte_write: 0,
+        });
+    }
+
+    fun validate_usage_config(config: &UsageGasConfig) {
+        assert!(config.target_usage > 0, error::invalid_argument(EZERO_TARGET_USAGE));
+        assert!(config.target_usage <= 18446744073709551615 / BASIS_POINT_DENOMINATION, error::invalid_argument(ETARGET_USAGE_TOO_BIG));
+        validate_curve(&config.read_curve);
+        validate_curve(&config.create_curve);
+        validate_curve(&config.write_curve);
+    }
+
+    fun validate_curve(curve: &GasCurve) {
+        assert!(curve.max_gas >= curve.min_gas, error::invalid_argument(EINVALID_GAS_RANGE));
+        let points = &curve.points;
+        let len = vector::length(points);
+        assert!(len >= 2, error::invalid_argument(EINVALID_CURVE_LENGTH));
+        assert!(vector::borrow(points, 0).x > 0, error::invalid_argument(EINVALID_CURVE_START_POINT));
+        assert!(vector::borrow(points, len - 1).x < BASIS_POINT_DENOMINATION, error::invalid_argument(EINVALID_CURVE_END_POINT));
+        let i = 0;
+        while (i <= len) {
+            let cur = if (i == 0) { &Point {x: 0, y: 0} } else { vector::borrow(points, i - 1) };
+            let next = if (i == len) { &Point {x: BASIS_POINT_DENOMINATION, y: BASIS_POINT_DENOMINATION} } else { vector::borrow(points, i) };
+            assert!(cur.x < next.x && cur.y <= next.y, error::invalid_argument(EINVALID_MONOTONICALLY_NON_DECREASING_CURVE));
+            i = i + 1;
+        }
```

Comment:
Since you already checked the points at the two ends, you can just loop from i = 0 to (len - 2) inclusive and get rid of the special checks.

Reply:
I checked the points at the ends. but if doesn't tells me points[len - 1].x < end_point.x and points[len - 1].y <= end_point.y. So tho I have len points, actually we have len + 2 in total and have to compare len + 1 pairs that's why we loop i in [0, len], which has len + 1 iterations.